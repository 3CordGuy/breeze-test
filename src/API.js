const getGroups = () =>
    fetch(`//localhost:8000/api/groups/`).then((response) => response.json());

const getPeople = () =>
    fetch(`//localhost:8000/api/people/`).then((response) => response.json());

const removePerson = (person_id) =>
    fetch(`//localhost:8000/api/people/${person_id}`, {
        method: "DELETE",
        headers: {
            "Content-Type": "application/json",
        },
    });

const updatePerson = (person) =>
    fetch(`//localhost:8000/api/people/${person.id}`, {
        method: "PUT",
        headers: {
            "Content-Type": "application/json",
        },
        body: JSON.stringify(person),
    });

const importPeople = (people) =>
    fetch(`//localhost:8000/api/people-import`, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
        },
        body: JSON.stringify(people),
    });

const importGroups = (groups) =>
    fetch(`//localhost:8000/api/groups-import`, {
        method: "POST",
        mode: "cors",
        headers: {
            "Content-Type": "application/json",
        },
        body: JSON.stringify(groups),
    });

const getGroup = (group_id) =>
    fetch(`http://localhost:8000/api/groups/${group_id}`).then((response) =>
        response.json(),
    );

const API = {
    getGroups,
    getPeople,
    removePerson,
    updatePerson,
    importPeople,
    importGroups,
    getGroup,
};

export default API;
