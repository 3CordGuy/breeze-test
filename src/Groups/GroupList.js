import React, { Component } from "react";
import { Table, Header, Dimmer, Loader, Icon } from "semantic-ui-react";
import { Link } from "@reach/router";
import _ from "lodash";

class GroupList extends Component {
    constructor(props) {
        super(props);
        this.state = {
            loading: true,
            data: [],
            column: "name",
            direction: "descending",
        };
    }

    handleSort = (clickedColumn) => () => {
        const { column, data, direction } = this.state;

        if (column !== clickedColumn) {
            this.setState({
                column: clickedColumn,
                data: _.sortBy(data, [clickedColumn]),
                direction: "ascending",
            });

            return;
        }

        this.setState({
            data: data.reverse(),
            direction: direction === "ascending" ? "descending" : "ascending",
        });
    };

    componentDidMount() {
        fetch("http://localhost:8000/api/groups")
            .then((response) => response.json())
            .then(({ data }) =>
                this.setState({
                    loading: false,
                    data: _.sortBy(data, ["name", "descending"]),
                }),
            );
    }

    render() {
        let data = this.state.data || [];
        let { column, direction, loading } = this.state;

        return (
            <>
                {loading && (
                    <Dimmer active inverted>
                        <Loader>Loading Data...</Loader>
                    </Dimmer>
                )}
                <Header as="h1">
                    <Icon name="users" />
                    Groups
                </Header>
                <Table celled padded basic="very" sortable>
                    <Table.Header>
                        <Table.Row>
                            <Table.HeaderCell
                                singleLine
                                sorted={column === "name" ? direction : null}
                                onClick={this.handleSort("name")}
                            >
                                Name
                            </Table.HeaderCell>
                        </Table.Row>
                    </Table.Header>

                    <Table.Body>
                        {data.map((group, index) => {
                            return (
                                <Table.Row key={index}>
                                    <Table.Cell singleLine>
                                        <Link to={`${group.id}`}>
                                            {group.name}
                                        </Link>
                                    </Table.Cell>
                                </Table.Row>
                            );
                        })}
                    </Table.Body>
                </Table>
            </>
        );
    }
}

export default GroupList;
