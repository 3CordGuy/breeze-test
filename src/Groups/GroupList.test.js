import React from "react";
import { shallow, mount } from "enzyme";
import GroupList from "./GroupList";
import API from "../API";
import _ from "lodash";

const GROUP_LIST = {
    data: [
        {
            id: 62,
            name: "Pawnee Book Club",
            created_at: "2020-06-28T01:17:03.000000Z",
            updated_at: "2020-06-28T01:17:03.000000Z",
        },
        {
            id: 63,
            name: "Dunder Mifflin Fan Club",
            created_at: "2020-06-28T01:17:03.000000Z",
            updated_at: "2020-06-28T01:17:03.000000Z",
        },
    ],
    total: 2,
};

let wrapper, data;

// SNAPSHOT TEST
describe("<GroupList />", () => {
    beforeAll(() => {
        wrapper = shallow(<GroupList />);
        data = [
            {
                id: 132,
                name: "Awesome Group",
                updated_at: "2019-07-20 22:05:47",
                created_at: "2019-07-20 22:05:47",
            },
        ];

        wrapper.setState({ data: data });
    });

    test("should match the snapshot", () => {
        expect(wrapper).toMatchSnapshot();
    });
});

// // UNIT TESTS
describe("<GroupList />", () => {
    it("should show a proper loader", () => {
        const wrapper = mount(<GroupList />);
        wrapper.setState({ loading: true });

        let loader = wrapper.find("div.loader");

        expect(loader.text()).toBe("Loading Data...");
    });

    it("should fetch the groups async", () => {
        const promise = new Promise((resolve) => resolve(GROUP_LIST));

        API.getGroups = jest.fn(() => promise);

        const wrapper = mount(<GroupList />);

        expect(wrapper.find("table tbody tr").length).toEqual(1);

        promise.then(() => {
            setImmediate(() => {
                wrapper.update();
                expect(wrapper.find("table tbody tr").length).toEqual(2);

                API.getGroups.mockClear();
            });
        });
    });

    it("should sort list by group name", () => {
        const wrapper = mount(<GroupList />);
        wrapper.setState({
            direction: "asc",
            data: _.orderBy(GROUP_LIST.data, "name", ["asc"]),
        });

        let nameHeader = wrapper.find("th.ascending.sorted");

        nameHeader.simulate("click");
        wrapper.update();

        expect(wrapper.state().data).toEqual(
            _.orderBy(GROUP_LIST.data, "name", ["desc"]),
        );

        nameHeader.simulate("click");

        expect(wrapper.state().data).toEqual(
            _.orderBy(wrapper.state().data, "name", ["asc"]),
        );
    });
});
