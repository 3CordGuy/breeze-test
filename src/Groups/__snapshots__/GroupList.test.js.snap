// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GroupList /> should match the snapshot 1`] = `
<Fragment>
  <Header
    as="h1"
  >
    Groups
  </Header>
  <Table
    as="table"
    celled={true}
    padded={true}
    selectable={true}
    sortable={true}
  >
    <TableHeader
      as="thead"
    >
      <TableRow
        as="tr"
        cellAs="td"
      >
        <TableHeaderCell
          as="th"
          onClick={[Function]}
          singleLine={true}
          sorted="descending"
        >
          Name
        </TableHeaderCell>
      </TableRow>
    </TableHeader>
    <TableBody
      as="tbody"
    >
      <TableRow
        as="tr"
        cellAs="td"
        key="0"
      >
        <TableCell
          as="td"
          singleLine={true}
        >
          <Link
            to="group/132"
          >
            Awesome Group
          </Link>
        </TableCell>
      </TableRow>
    </TableBody>
  </Table>
</Fragment>
`;
